### PR TITLE
Issue #55: Unit tests generated, and python code to generate these tests

### PR DESCRIPTION
**Pull request by AlexandreFournel**

### CODE REVIEW

**Comment by cyrille-artho** [comment]
```
/home/travis/build/cyrille-artho/modbat/src/test/scala/modbat/config/ModbatConfig.scala:1: not found: type Configuration
class TraitImpl extends Configuration 
```
Maybe an import statement is missing?

**Comment by cyrille-artho** [comment]
Two compile errors remain:
```
/home/travis/build/cyrille-artho/modbat/src/test/scala/modbat/mbt/ConfigTest.scala:8: value _1 is not a member of Unit
    result._1 should be(0)
           ^
/home/travis/build/cyrille-artho/modbat/src/test/scala/modbat/mbt/ConfigTest.scala:9: value _3 is not a member of Unit
    result._3 shouldBe empty
           ^
two errors found
```

**Comment by cyrille-artho** [comment]
Still a lot of failing tests. Please look into them.
The tests where logging fails due to "no such file or directory" can be fixed by extending the test harness, so it automatically creates the missing directories if they do not exist.

**Comment by AlexandreFournel** [comment]
Alright, 
What do you mean about extend the test harness ? Do i have to change the dependencies of files ? 



Alexandre FOURNEL 



Élève ingénieur 2ème année 
[ https://www.ensta-paristech.fr/?utm-source=mails ] 	


ENSTA ParisTech 

828, boulevard des Maréchaux 

91120 PALAISEAU [ [ https://www.ensta-paristech.fr/fr/venir-ensta-paristech | plan ] ] 

[ https://www.ensta-paristech.fr/ | www.ensta-paristech.fr ] 


Nos réseaux : [ https://www.facebook.com/ENSTA.ParisTech | Facebook ] | [ https://www.twitter.com/ENSTAParisTech | Twitter ] | [ https://www.instagram.com/enstaparistech/ | Instagram ] | [ https://www.youtube.com/user/ENSTAParisTech | YouTube ] | [ https://www.linkedin.com/company/ensta | LinkedIn ] 


De: "cyrille-artho" <notifications@github.com> 
À: "cyrille-artho/modbat" <modbat@noreply.github.com> 
Cc: "AlexandreFournel" <alexandre.fournel@ensta-paristech.fr>, "Author" <author@noreply.github.com> 
Envoyé: Mercredi 5 Juin 2019 12:00:06 
Objet: Re: [cyrille-artho/modbat] Issue #55: Unit tests generated, and python code to generate these tests (#64) 



Still a lot of failing tests. Please look into them. 
The tests where logging fails due to "no such file or directory" can be fixed by extending the test harness, so it automatically creates the missing directories if they do not exist.

File: src/test/scala/modbat/mbt/ModbatTestHarness.scala [hunk 1]
```diff
@@ -20,12 +20,38 @@ import modbat.log.Log
 import modbat.util.CloneableRandom
 
 object ModbatTestHarness {
+  def getNamesofFile(args: Array[String], td: org.scalatest.TestData): (String, String, String) = {
+    // Be careful, this function create folder if doesn't exist
```
Comment (cyrille-artho) [inline src/test/scala/modbat/mbt/ModbatTestHarness.scala]: Write a comment at the beginning of the function, describing its overall purpose, and the meaning of the return values. At the end, you can add
"This function creates a directory if it does not exist."

File: src/test/scala/modbat/mbt/ModbatTestHarness.scala [hunk 2]
```diff
@@ -20,12 +20,38 @@ import modbat.log.Log
...
+    var name_output = "log/modbat/"
+    var arguments_name = ""
+    for ( x <- args ) {
+      if (x contains "modbat"){
```
Comment (cyrille-artho) [inline src/test/scala/modbat/mbt/ModbatTestHarness.scala]: Perhaps this should be "modbat.", as part of a package name, so the substring match is more specific?

File: src/test/scala/modbat/mbt/ModbatTestHarness.scala [hunk 3]
```diff
@@ -20,12 +20,38 @@ import modbat.log.Log
...
+    (name_output, arguments_name, ( td.name.split(" ") )(0))
+  }
 
-  def using[A <: {def close(): Unit}, B](resource: A)(f: A => B): B =
-  try f(resource) finally resource.close()
-
-  def writeToFile(path: String, data: String): Unit = 
-    using(new FileWriter(path))(_.write(data))
+  def addToFiles(writer_1: java.io.PrintWriter,writer_2: java.io.PrintWriter,data:String){
+    writer_1.write(data+"\n")
```
Comment (cyrille-artho) [inline src/test/scala/modbat/mbt/ModbatTestHarness.scala]: Use "println" here and in the next line.

File: src/test/scala/modbat/mbt/ModbatTestHarness.scala [hunk 4]
```diff
@@ -35,8 +61,7 @@ object ModbatTestHarness {
   }
 
   def compare_iterators(it1:Iterator[String], it2:Iterator[String]) : Boolean = {
-    var current_iterator="There is no element in the Iterator."
-    var current_iterator2="There is no element in the Iterator."
+    var current_iterator, current_iterator2="There is no element in the Iterator."
```
Comment (cyrille-artho) [inline src/test/scala/modbat/mbt/ModbatTestHarness.scala]: Use only *one* "current" (past) iterator (for the case when the lines match); change the text to "End of file."

File: src/test/scala/modbat/mbt/ModbatTestHarness.scala [hunk 5]
```diff
@@ -45,7 +70,7 @@ object ModbatTestHarness {
           current_iterator2 = it2.next
           if (current_iterator != current_iterator2){
             current_comparison=false
-            current_iterator="iterators different : \n validated_out_lines = " + current_iterator + "\n out = " + current_iterator2
+            current_iterator="iterators different : \n validated_out_lines = \n" + current_iterator + "\n out_filtered = \n" + current_iterator2
```
Comment (cyrille-artho) [inline src/test/scala/modbat/mbt/ModbatTestHarness.scala]: Print the message here, so you can still access the contents of both iterators, as well as the content of the previous line that matched.

File: src/test/scala/modbat/mbt/ModbatTestHarness.scala [hunk 6]
```diff
@@ -71,45 +96,36 @@ object ModbatTestHarness {
     current_comparison
   }
 
-  def filtering_and_regex(name_output : String, typeoutput : String, typeByte : ByteArrayOutputStream, name_file: String){
-    var regex_list = List("")
+  def filtering_writing_comparing_each_file(name_output_1 : String, name_output_2 : String, typeByte : ByteArrayOutputStream){
+    var regex_map = Map[String, String]()
```
Comment (cyrille-artho) [inline src/test/scala/modbat/mbt/ModbatTestHarness.scala]: The long name suggests that it is better to break up this function into two parts.
Use "camel case" notation, also in the functions before, e.g., "compareIterators".

File: src/test/scala/modbat/mbt/ModbatTestHarness.scala [hunk 7]
```diff
@@ -71,45 +96,36 @@ object ModbatTestHarness {
...
     var replace_sentences = List("")
-    if (typeoutput == "out"){
-      regex_list = List("""\[[0-9][0-9]*[mK]""", """.*""", """ in .*[0-9]* milliseconds""", """RELEASE-([0-9.]*)""",  """ v[0-9a-f]* rev [0-9a-f]*""", """ v[0-9][0-9]*\\.[0-9][^ ]* rev [0-9a-f]*""", """^Time: [0-9.]*""", """(at .*):[0-9]*:""", """canceled 0, """, """(\[INFO\] .* java.lang.AssertionError.*):""")
-      replace_sentences=List("", "", "", "$1", "", " vx.yz", " vx.yz", "$1", "", "$1")
+    if (name_output_1.split("\\.").last == "out"){
+      regex_map = Map("""\[[0-9][0-9]*[mK]"""->"",""".*"""->"",""" in .*[0-9]* milliseconds"""->"","""RELEASE-([0-9.]*)"""->"$1",""" v[0-9a-f]* rev [0-9a-f]*"""->"",""" v[0-9][0-9]*\\.[0-9][^ ]* rev [0-9a-f]*"""->" vx.yz","""^Time: [0-9.]*"""->" vx.yz","""(at .*):[0-9]*:"""->"$1","""canceled 0, """->"","""(\[INFO\] .* java.lang.AssertionError.*):"""->"$1")
```
Comment (cyrille-artho) [inline src/test/scala/modbat/mbt/ModbatTestHarness.scala]: Use line breaks to put each pair on a new line.

File: src/test/scala/modbat/mbt/ModbatTestHarness.scala [hunk 8]
```diff
@@ -214,23 +196,22 @@ object ModbatTestHarness {
...
   def setFooJar() = {
-    val mapsetTestJar = new java.util.HashMap[String, String]()
-    mapsetTestJar.put("CLASSPATH", "foo")
-    setEnv (mapsetTestJar)
+    setJar("foo")
   }
 
-}
-
+}
```
Comment (cyrille-artho) [inline src/test/scala/modbat/mbt/ModbatTestHarness.scala]: A file should always have a newline character at the end.

File: src/test/scala/modbat/mbt/ModbatTestHarness.scala (hunk 6 above)
Comment (cyrille-artho) [reply src/test/scala/modbat/mbt/ModbatTestHarness.scala]: Also, make sure that the validated template that is *read* has extension .out or .eout, but the files that are *written* with the current output end in .log or .err, respectively.

File: src/test/scala/modbat/mbt/ModbatTestHarness.scala [hunk 9]
```diff
@@ -71,45 +96,36 @@ object ModbatTestHarness {
...
-    else{
-      sentence
-    }
+  def filtering_writing_comparing(name_output: String, arguments_name: String, simple_name:String, out : ByteArrayOutputStream, err : ByteArrayOutputStream){
+    filtering_writing_comparing_each_file (name_output+arguments_name+".out", name_output+simple_name+".out", out)
+    filtering_writing_comparing_each_file (name_output+arguments_name+".eout", name_output+simple_name+".eout", err)
+  }
+
+  def replaceRegexInSentence(sentence : String, regex_map : Map[String,String]) : String = {  
+    var sentence_filtred = sentence
```
Comment (cyrille-artho) [inline src/test/scala/modbat/mbt/ModbatTestHarness.scala]: spelling

**Comment by cyrille-artho** [comment]
This PR is outdated and cannot be merged as is due to too many merge conflicts.
Instead, individual refactorings should be applied incrementally, once other branches with work in progress are merged and closed.

**Pull request closed without merging**